pkg/csi/service/driver.go

```diff
@@ -134,6 +140,54 @@ func (driver *vsphereCSIDriver) BeforeServe(ctx context.Context) error {
 			log.Errorf("failed to read config. Error: %+v", err)
 			return err
 		}
+
+		if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.CSIInternalGeneratedClusterID) {
+			CSINamespace := common.GetCSINamespace()
+
+			if clusterFlavor == cnstypes.CnsClusterFlavorVanilla && cfg.Global.ClusterID == "" {
+				// In case of vanilla k8s deployments, if cluster ID is not provided in the
+				// vSphere config secret, then generate an unique cluster ID internally.
+				clusterID := getNewUUID()
+
+				cmData, err := commonco.ContainerOrchestratorUtility.GetConfigMap(ctx,
+					cnsconfig.ClusterIDConfigMapName, CSINamespace)
+				if err == nil {
+					// If ConfigMap for cluster ID already exists, then instead of
+					// using newly generated clusterID value, we will use the
+					// clusterID value stored in the existing immutable ConfigMap.
+					clusterID = cmData["clusterID"]
+					log.Infof("clusterID is not provided in vSphere Config Secret, "+
+						"using the clusterID %s from existing ConfigMap", clusterID)
+				} else {
+					// Create the immutable ConfigMap to store cluster ID, so that it will be
+					// persisted in etcd and it can't be updated by any user.
+					configMapData := map[string]string{"clusterID": clusterID}
+
+					err := commonco.ContainerOrchestratorUtility.CreateConfigMap(ctx,
+						cnsconfig.ClusterIDConfigMapName, CSINamespace, configMapData, true)
+					if err != nil {
+						return logger.LogNewErrorf(log, "Failed to create the immutable ConfigMap, Err: %v",
+							err)
+					}
+					log.Infof("clusterID is not provided in vSphere Config Secret, "+
+						"generated a new clusterID %s", clusterID)
+				}
+
+				cnsconfig.GeneratedClusterID = clusterID
```

Comment:
This will not be used by the syncer container

Reply:
I have added wrapper functions in syncer container to get the correct cluster ID.